# Assignment for week 5

Use the following table to provide us with

|name | exam number|
|----|----|
|Kevin van Dijk| u1268889|
|David Dixon| u1278247|

In the lecture we considered a two period bargaining model. In this assignment we are going to extend this to 3 periods.

Pay attention to what you do here. The python part of the final assignment will be the extension to $n$ bargaining periods; but let us not worry about this yet.

We need the following library:

In [2]:
import numpy as np

As we are looking at more periods here, we need a finer grid of offers. We want all offers between 0.0 and 10.0 (including 10.0) with step 0.1 between offers.

**1)** Define the list `offers` using `np.arrange`:

In [3]:
min_offer = 0.0
max_offer = 10.0
step = 0.1
offers = np.arange(min_offer, max_offer+0.1,0.1)
offers

array([  0. ,   0.1,   0.2,   0.3,   0.4,   0.5,   0.6,   0.7,   0.8,
         0.9,   1. ,   1.1,   1.2,   1.3,   1.4,   1.5,   1.6,   1.7,
         1.8,   1.9,   2. ,   2.1,   2.2,   2.3,   2.4,   2.5,   2.6,
         2.7,   2.8,   2.9,   3. ,   3.1,   3.2,   3.3,   3.4,   3.5,
         3.6,   3.7,   3.8,   3.9,   4. ,   4.1,   4.2,   4.3,   4.4,
         4.5,   4.6,   4.7,   4.8,   4.9,   5. ,   5.1,   5.2,   5.3,
         5.4,   5.5,   5.6,   5.7,   5.8,   5.9,   6. ,   6.1,   6.2,
         6.3,   6.4,   6.5,   6.6,   6.7,   6.8,   6.9,   7. ,   7.1,
         7.2,   7.3,   7.4,   7.5,   7.6,   7.7,   7.8,   7.9,   8. ,
         8.1,   8.2,   8.3,   8.4,   8.5,   8.6,   8.7,   8.8,   8.9,
         9. ,   9.1,   9.2,   9.3,   9.4,   9.5,   9.6,   9.7,   9.8,
         9.9,  10. ])

**2)** Copy the function `accept_offer` from the lecture:

In [4]:
def accept_offer(offer,your_outside_option):
    accept = (offer >= your_outside_option)
    return accept

**3)** Copy the function `make_offer` from the lecture:

In [5]:
def make_offer(your_outside_option,other_outside_option):
    profits = [(max_offer-offer)*accept_offer(offer,other_outside_option) for offer in offers]
    max_profit = max(profits)
    max_index = profits.index(max_profit)
    if max_profit >= your_outside_option:
        your_offer = offers[max_index]
        your_profit = max_offer-offers[max_index]
        other_profit = offers[max_index]
    else:
        your_offer = -1 # no offer is made
        your_profit = your_outside_option
        other_profit = other_outside_option
    return your_offer, your_profit, other_profit

Pay attention to what `make_offer` returns and in which order. You need this below.

In this assignment, we are looking at the three period bargaining problem. We want to use the function `make_offer` to keep track of the outside options without defining a new dictionary `outside_options` (as we did in the lecture).

As in the lecture, we will use backward induction to solve this. That is, we first consider what happens in period 3 --in case the offers made in periods 1 and 2 are rejected. Once we know what will happen in period 3, we can move back to period 2 and then to period 1.

To do this, we create a dictionary `offer`. 

**4)** Explain what the following code does (e.g. what is `offer[4]`?).

In [6]:
delta = 0.9
offer = {}
offer[4] = [0,0,0]
offer[3] = make_offer(delta*offer[4][2],delta*offer[4][1])



Delta = 0.9 represents the discount factor after each round. 

Offer[4] is the outcome after the 3 period model game. Where the offer is your_offer = 0 and your_profit = 0 but also the other_profit = 0.

The game is basically over and each moves on without any money. 

**5)** Using the logic above, define `offer[2]` and `offer[1]`:

In [7]:
offer[2] = make_offer(delta*offer[3][2],delta*offer[3][1])
offer[1] = make_offer(delta*offer[2][2],delta*offer[2][1])

**6)** What are the payoffs for players 1 and 2 in a three period version of the bargaining model in the lecture?
</br>

{1: (0.90000000000000002, 9.0999999999999996,0.90000000000000002),</br>
 
 2: (9.0, 1.0, 9.0),</br>
 
 3: (0.0, 10.0, 0.0),
 
 4: [0, 0, 0]}</br>

This is the payout in each period. Period 4 signifies that there are no payouts if there isn't made a deal within 3 periods. 


**7)** In which period will the players reach an agreement?

In period 1.